### PR TITLE
Proxy configuration

### PR DESCRIPTION
**Pull request by carlosmn**
Determining the proxy from the config is but one way we can do it.

This would depend on having nicer option structures, so proper binding-in would need #3066 to go in, but we can figure out what the interface should be beforehand.

Part of the idea is to ask the caller for the proxy's pasword the same way we do for the git server's. We can re-use the credentials callbacks and structs to perform an interactive request to an end-user.

The code for WinHTTP is completely untested, but it *should* be in the right direction (according to the documentation) for grabbing the username/password from the config and using that. This was my first approach at solving #2106.

We would also need to think about what we want to do for nicer higher-level abstractions, since in e.g. pygit2[0] you can currently provide a callable object as a callback. It would be nice if we could make it possible to provide one of these objects for each if you know beforehand what they are, but don't want to put them in urls or write them in the plain-text config.

[0] I think libgit2sharp may also provide this, but the only time I used it in anger I rewrote the network stack.

- [x] Work with WinHTTP
- [x] Work with libcurl
- [x] Don't require a trailing slash for the hostname
- [x] Remember an accepted proxy credential object

### CODE REVIEW

**Comment by carlosmn** [comment]
The Windows bits for this would depend on having some proxy we can use on AppVeyor, @ethomson had mentioned he had some program we wanted to use for this.

On the nixes, we should be able to use OpenSSH to create a SOCKS proxy. For HTTP proxies, something like tinyproxy looks promising.

File: include/git2/proxy.h [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+	/**
+	 * The URL of the proxy.
+	 */
+	const char *url;
+
+	/**
+	 * The port of the proxy
+	 */
+	const char *port;
```
Comment (shiftkey) [inline include/git2/proxy.h]: I can't see this used anywhere in the PR. Feels?

Comment (carlosmn) [inline include/git2/proxy.h]: We should probably delete it and just have the url for the whole thing. When we use our own connection functions, we split it out, but IIRC both WinHTTP and curl will happily take it as a url string.

**Comment by carlosmn** [comment]
This code currently builds and runs on Windows (and hopefully nothing else is broken), both with asking the credentials from the user as well as getting them from the url, which would also take care of #2106.

On unix, using the curl env vars should also magically take care of this, but we also need to hook this in.

File: appveyor.yml [hunk 2]
```diff
@@ -32,8 +33,17 @@ build_script:
...
       cmake -D ENABLE_TRACE=ON -D BUILD_CLAR=ON -D MSVC_CRTDBG=ON .. -G"$env:GENERATOR"
       cmake --build . --config Debug
+      Invoke-WebRequest https://github.com/ethomson/poxyproxy/releases/download/v0.1.0/poxyproxy-0.1.0.jar -OutFile poxyproxy.jar
     }
 - cmd: |
     if "%GENERATOR%"=="MSYS Makefiles" (C:\MinGW\msys\1.0\bin\sh --login /c/projects/libgit2/script/appveyor-mingw.sh)
 test_script:
-- ps: ctest -V .
+- ps: |
+    Start-Job { java -jar C:\projects\libgit2\build\poxyproxy.jar -d --port 8080 --credentials foo:bar }
```
Comment (nulltoken) [inline appveyor.yml]: Maybe use `java -jar $Env:APPVEYOR_BUILD_FOLDER\build\poxyproxy.jar` ?

File: appveyor.yml [hunk 3]
```diff
@@ -32,8 +33,17 @@ build_script:
...
       cmake -D ENABLE_TRACE=ON -D BUILD_CLAR=ON -D MSVC_CRTDBG=ON .. -G"$env:GENERATOR"
       cmake --build . --config Debug
+      Invoke-WebRequest https://github.com/ethomson/poxyproxy/releases/download/v0.1.0/poxyproxy-0.1.0.jar -OutFile poxyproxy.jar
     }
 - cmd: |
     if "%GENERATOR%"=="MSYS Makefiles" (C:\MinGW\msys\1.0\bin\sh --login /c/projects/libgit2/script/appveyor-mingw.sh)
 test_script:
-- ps: ctest -V .
+- ps: |
+    Start-Job { java -jar $Env:APPVEYOR_BUILD_FOLDE\build\poxyproxy.jar -d --port 8080 --credentials foo:bar }
```
Comment (nulltoken) [inline appveyor.yml]: I think you're missing a trailing `R` in the env name

File: script/cibuild.sh [hunk 4]
```diff
@@ -49,10 +49,24 @@ export GITTEST_REMOTE_SSH_KEY="$HOME/.ssh/id_rsa"
 export GITTEST_REMOTE_SSH_PUBKEY="$HOME/.ssh/id_rsa.pub"
 export GITTEST_REMOTE_SSH_PASSPHRASE=""
 
+
+# Can we ask Travis to cache this file?
+curl https://github.com/ethomson/poxyproxy/releases/download/v0.1.0/poxyproxy-0.1.0.jar >poxyproxy.jar || exit $?
```
Comment (arthurschreiber) [inline script/cibuild.sh]: You could put it somewhere like `vendor/poxyproxy/poxyproxy.jar` and then add this to the travis config:

```
cache:
  directories:
  - vendor/poxyproxy
```

**Comment by carlosmn** [comment]
I'm thinking that we should remove the `proxy_type` completely and just rely on the URL to tell us what kind of proxy you want.

**Comment by bso-intel** [comment]
Is this still in progress?
Does libgit2 support proxy yet?

**Comment by carlosmn** [comment]
@bso-intel libgit2 supports proxies the same way git does: you tell curl what to use as a proxy

**Comment by bso-intel** [comment]
Great!! Thank you very much, Carlos.
I have been waiting for this proxy support for a year!!

**Comment by ethomson** [comment]
> Great!! Thank you very much, Carlos.
I have been waiting for this proxy support for a year!!

@bso-intel To clarify - we've had proxy support for quite a long time.  It's not new in this pull request.  Are you having trouble using it?

**Comment by bso-intel** [comment]
@ethomson 
I was actually wanted to use nodegit which is a node export of libgit2.
They say they do not support proxy because libgit2 does not support it.
I actually tried nodegit about a year ago, and found that nodegit does not support proxy on Mac.
I needed proxy support for windows, Mac and Linux.
I hope this pull request solve that proxy issue on all these platforms.

**Comment by ethomson** [comment]
If you're using nodegit, please open an issue in their repository.  Feel free to cc me (just add `cc @ethomson` to the issue message) and I can help!

**Comment by tbranyen** [comment]
@ethomson sorry very confused by your suggestion to open a PR in NodeGit. Are you saying that libgit2 supports an option we can enable in NodeGit to allow the passing of a proxy URL? If not, opening an issue in our project will not be helpful since it's out of the scope.

**Comment by ethomson** [comment]
@tbranyen We support `remote.<remote>.proxy` and environment variables.  If there's something else that your users request, it probably makes sense for them to ask nodegit for it - or at least to start the conversation there.  Even if we add support for something in libgit2, it's not clear that would satisfy an end user of a binding.

**Comment by tbranyen** [comment]
Ah okay, thanks for clearing that up @ethomson. Maybe this is something @bso-intel can spearhead, since it seems all we need is a unit test/documentation to show how this works.

**Comment by johnhaley81** [comment]
I think there was some confusion going on with https://github.com/nodegit/nodegit/issues/489. The underlying libgit2 issue (https://github.com/libgit2/libgit2/issues/2555) I think we mis-interpreted as implementing proxy support in libgit2 _at all_ outside of windows which is not the case.

**Comment by bso-intel** [comment]
So, it seems libgit2 still does not support proxy setting for Mac and Linux yet.
I cannot rely on setting the env var for proxy setting.
I need to programmatically set the proxy from a node app on windows, mac, and Linux.
Did libgit2 ever plan to support this sometime in the future?

**Comment by tbranyen** [comment]
@bso-intel why can't you modify `process.env` dynamically?

**Comment by bso-intel** [comment]
I have never thought about that. Great idea. Thanks, Tim.

**Comment by cbargren** [comment]
Hey @carlosmn, what's the status of this PR? Is the checklist of things to be done up to date?

**Comment by carlosmn** [comment]
I've got the appveyor tests passing for MSVC but when running the mingw tests, it doesn't seem to be running.

**Comment by carlosmn** [comment]
CI is clean now, so I think it's ready for merging. I think we should just merge it and figure out if there's missing features later when we wrap it into the bindings and people start using them. @ethomson thoughts?

**Comment by dg-ratiodata** [comment]
@carlosmn Great! Is there documentation about how to use this anywhere? I had a look at the changed code, but didn't found a hint for that. But anyway, thanks for adding this feature to libgit2

**Comment by carlosmn** [comment]
It's documented the same way as the rest of the options you can use for fetching or pushing. I'll update the generated docs so this gets rendered on the reference webpage.

**Comment by dg-ratiodata** [comment]
Thanks.

**Pull request merged**